Add LitElement basics article (#245)
vaadin/docs: articles/fusion/application/lit-element.asciidoc

LitElement provides a simple base class, and a set of tools to create reactive web components.
Under the hood, it uses link:https://lit-html.polymer-project.org[lit-html], a tiny library to render and re-render HTML templates in the most efficient and declarative way.

== Minimal LitElement Component

=== Templates

A class that inherits the LitElement base class must implement a method called `render()` which returns a `TemplateResult` instance produced by the `html` tag.
The string tagged with the `html` tag is displayed in the browser as the component's content.

More about tags and tagged template literals you can find in link:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates[MDN documentation].

=== Custom Element Definition

A class that pretends to be a web component must be added to the component registry and assigned with the HTML element name.
With LitElement, you may do it with the `@customElement` decorator.
It receives an HTML element name which should contain two parts split by a dash `-`. For example, `my-view`, `x-button` or `vaadin-checkbox`.

> 

Reviewer: at least one dash?

I think multiple dashes is also fine.

=== Example

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/templates.ts[]
----

== Data Binding

=== Fields, Methods, and Accessors

Since the web component is a simple JavaScript class, you may define any number of fields, methods, and accessors you need and call them as you are used to.
Typescript as the JavaScript superset also allows defining private and protected methods.

> 

Reviewer: Could have an example to show a field, a method, private field, private method etc?

Author: Yes, I put the common example at the end of the Data Binding section which also includes a `private _onClick()` method

----

== Shadow and Light DOM

You may use two main approaches for web components.
You can find additional information in the <<{articles}/ds/customization/style-scopes#,Style Scopes>> article.

=== Shadow DOM
A web component can create a shadow tree within.
It is a separate document fragment almost inaccessible from the regular DOM existing as a part of the component and displayed in the browser as a regular element tree.
With it, you can create elements with styles independent of the global CSS and HTML content not added to children nodes.

Shadow DOM is enabled by default in LitElement.

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/shadow-dom.ts[]
----

=== Light DOM

If you don't want to use the Shadow DOM you may still use the Light DOM, a regular part of the DOM controlled by the component's class.
It lacks Shadow DOM advantages like scoped CSS or non-children HTML but it is lighter and simpler for both development and performance.

To enable Light DOM for the LitElement component, add a `createRenderRoot() { return this; }` method to your component.

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/light-dom.ts[]
----

> === Usage Advices

Reviewer 2: Plural is the same as singular for advice.
```suggestion
=== Usage Advice
```

Author: Fixed

---
title: LitElement Basics
order: 20
layout: page
---

= LitElement Basics

For creating views on the frontend side, Vaadin Fusion uses web components, a native technology that allows creating custom HTML elements.
It is almost like a framework built into the browser, so in most cases it is faster and smaller than any other framework, while still shares their advantages.

However, since the in-browser implementation of the web components is low-level, to make them more convenient, it is recommended to use a LitElement library.

LitElement provides a simple base class, and a set of tools to create reactive web components.
Under the hood, it uses link:https://lit-html.polymer-project.org[lit-html], a tiny library to render and re-render HTML templates in the most efficient and declarative way.

== Minimal LitElement Component

> 

Author: Fixed

=== Templates

A class that inherits the LitElement base class must implement a method called `render()` which returns a `TemplateResult` instance produced by the `html` tag.
The string tagged with the `html` tag is displayed in the browser as the component's content.

More about tags and tagged template literals you can find in link:https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates[MDN documentation].

=== Custom Element Definition

A class that pretends to be a web component must be added to the component registry and assigned with the HTML element name.
With LitElement, you may do it with the `@customElement` decorator.
It receives an HTML element name which should contain at least one dash `-`.
For example, `my-view`, `x-button` or `vaadin-checkbox`.

=== Example

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/minimal.ts[]
----

== Data Binding

> 

Author: Fixed

For that approach, you must add a question mark `?` before the property name.
* Listen to an event by the name.
Allows only functions.
When the element fires an event, the function assigned to the name is called with the event as an argument.
For that approach, you must add an at sign `@` before the event name: `@event`.

**NOTE**: You don't need to bind `this` to the event listener method.
LitElement calls any listener with the correct `this` by design.

More about how to write templates in the link:https://lit-element.polymer-project.org/guide/templates[LitElement documentation].

=== Example

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/data-binding.ts[]
----

== Lifecycle

Each component has lifecycle methods that may be overridden to execute some code at a specific time.
To the standard web component lifecycle, LitElement adds its own methods to handle the asynchronous `render` calls.

Here is the list of LitElement lifecycle methods.
More about them you can read in the link:https://lit-element.polymer-project.org/guide/lifecycle[LitElement documentation].

> * `constructor`
> 
> **Specificity**: JavaScript class
> 
> The constructor is invoked when a class is instantiated.
> For a component, it happens when `document.createElement('my-view')` is called.
> 
> In most cases, you don't need a `constructor` because LitElement handles everything important.
> However, if you do, try to avoid performing any heavy work: the element might be discarded after invoking the constructor.
> 
> **NOTE**: when the element is created, no render happens.
> The element needs to be connected to run the `render` callback.
> 
> * `connectedCallback`
> 
> **Specificity**: Web Components
> 
> This callback is invoked each time the element is connected to the DOM.
> 
> In most cases, the `connectedCallback` is used for setting up component-level event listeners, etc.
> 
> **NOTE**: since `render` is asynchronous it does not over when the `connectedCallback` finishes.
> Use <<first-updated,`firstUpdated`>> callback to execute the code after the first render.
> 
> **NOTE**: you must call `super.connectedCallback()` in your `connectedCallback` method.
> Otherwise, you may lose the default LitElement lifecycle.
> 
> * `disconnectedCallback`
> 
> **Specificity**: Web Components
> 
> This callback is invoked each time the element is disconnected from the DOM.
> 
> The default role of the `disconnectedCallback` is to remove or close everything that is set up during the `connectedCallback`.
> 
> **NOTE**: this callback is invoked __after__ the element is disconnected from the DOM.
> You do not have access to the DOM from within the component in this callback.
> 
> **NOTE**: you must call `super.disconnectedCallback()` in your `disconnectedCallback` method.
> Otherwise, you may lose the default LitElement lifecycle.
> 
> * `firstUpdated` [[first-updated]]
> 
> **Specificity**: LitElement
> 
> This callback is invoked right after the first render is over and before the first `updated` callback.
> It receives an object that contains all changed properties.
> 
> * `updated`
> 
> **Specificity**: LitElement
> 
> This callback is invoked after each render; for the first render, it is invoked right after the `firstUpdated` callback.
> Just like `firstUpdate`, it receives an object that contains all properties changed after the previous render.

Reviewer 2: - Use a definition list
- Proper admonitions with titles
- Cleanup Asciidoc formatting (single * and _)

```suggestion
.List of lifecycle callbacks
`constructor`::
+
*Specificity*: JavaScript class
+
The constructor is invoked when a class is instantiated.
For a component, it happens when `document.createElement('my-view')` is called.
+
In most cases, you don't need a `constructor` because LitElement handles everything important.
However, if you do, try to avoid performing any heavy work: the element might be discarded after invoking the constructor.
+
.The element needs to be connected to run the `render` callback.
[NOTE]
When the element is created, no render happens.

`connectedCallback`::
+
*Specificity*: Web Components
+
This callback is invoked each time the element is connected to the DOM.
+
In most cases, the `connectedCallback` is used for setting up component-level event listeners, etc.
+
.Use <<first-updated,`first-updated`>> callback to execute the code after the first render
[NOTE]
Since `render` is asynchronous it does not over when the `connectedCallback` finishes.
+
.Remember to call the `super.connectedCallback()`
[NOTE]
You must call `super.connectedCallback()` in your `connectedCallback` method.
Otherwise, you may lose the default LitElement lifecycle.

`disconnectedCallback`::
+
*Specificity*: Web Components
+
This callback is invoked each time the element is disconnected from the DOM.
+
The default role of the `disconnectedCallback` is to remove or close everything that is set up during the `connectedCallback`.
+
.No DOM access
[NOTE]
This callback is invoked _after_ the element is disconnected from the DOM.
+
.Remember to call the `super.disconnectedCallback()`
[NOTE]
You must call `super.disconnectedCallback()` in your `disconnectedCallback` method.
Otherwise, you may lose the default LitElement lifecycle.

[[first-updated]]`firstUpdated`::
+
*Specificity*: LitElement
+
This callback is invoked right after the first render is over and before the first `updated` callback.
It receives an object that contains all changed properties.

`updated`::
+
*Specificity*: LitElement
+
This callback is invoked after each render; for the first render, it is invoked right after the `firstUpdated` callback.
Just like `firstUpdate`, it receives an object that contains all properties changed after the previous render.
```

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/lifecycle.ts[]
----

== Styling

For styling, LitElement provides a static property `styles`.

> You may use either a stylesheet imported from the `.css` file or define it in the component directly using the `css` tagged function.

Reviewer 2: ```suggestion
You can use either a style sheet imported from a `.css` file or define an inline style sheet using the `css` tagged function.
```

> You may specify multiple styles for the `styles` property.
> Just use the array instead of a direct assigning.

Reviewer 2: ```suggestion
The `styles` property also accepts an array of style sheets.
```

----

== Shadow and Light DOM

You may use two main approaches for web components.
You can find additional information in the <<{articles}/ds/customization/style-scopes#,Style Scopes>> article.

=== Shadow DOM
A web component can create a shadow tree within.
It is a separate document fragment almost inaccessible from the regular DOM existing as a part of the component and displayed in the browser as a regular element tree.
With it, you can create elements with styles independent of the global CSS and HTML content not added to children nodes.

Shadow DOM is enabled by default in LitElement.

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/shadow-dom.ts[]
----

=== Light DOM

If you don't want to use the Shadow DOM you may still use the Light DOM, a regular part of the DOM controlled by the component's class.
It lacks Shadow DOM advantages like scoped CSS or non-children HTML but it is lighter and simpler for both development and performance.

To enable Light DOM for the LitElement component, add a `createRenderRoot() { return this; }` method to your component.

[source,typescript]
----
include::{root}/frontend/demo/fusion/lit-basics/light-dom.ts[]
----

> === Usage Advice

Reviewer 2: We use this title/heading in the Design System docs, and it feels appropriate here as well.

```suggestion
=== Best Practices
```

> It is recommended to use Shadow DOM only for design components (checkboxes, text fields, combo-boxes, etc.) and Light DOM for any other components like views or their parts.

Reviewer 2: ```suggestion
Shadow DOM is recommended only for "leaf" components (checkboxes, text fields, combo boxes, etc.) and Light DOM for other components like views or their parts.
```

> It would allow avoiding possible performance issues that might happen due to multiple nesting of shadow trees.

Reviewer 2: ```suggestion
Deeply nested shadow trees can impact performance negatively.
```

== Using in Routes

> About how to use components in the Vaadin Router you can read in the <<{articles}/fusion/routing/router#,Vaadin Router>> documentation.

Reviewer 2: ```suggestion
Read more about how to use components together with Vaadin Router in the <<{articles}/fusion/routing/router#,Vaadin Router>> documentation.
```

=== `firstUpdated` [[first-updated]]

**Specificity**: LitElement

This callback is invoked right after the first render is over and before the first `updated` callback. It receives an object that contains all changed properties.

=== `updated`

**Specificity**: LitElement

This callback is invoked after each render; for the first render, it is invoked right after the `firstUpdated` callback. Just like `firstUpdate`, it receives an object that contains all properties changed after the previous render.

=== Example

.my-view.ts
[source,typescript]
----
import {LitElement, html, customElement} from 'lit-element';

@customElement('my-button')
class MyButton extends LitElement {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
  }

  connectedCallback() {
    this.addEventListener('click', this.onClick);
  }

  disconnectedCallback() {
    this.removeEventListener('click', this.onClick);
  }

  render() {
    return html`<button>Some button</button>`;
  }

  private onClick() {
    console.log('I am clicked');
  }
}

export default MyView;
----

== LitElement templates

> The `render()` method should return a `TemplateResult` instance produced by the `html` tagged function. More about how to write templates in the link:https://lit-element.polymer-project.org/guide/templates[LitElement documentation].

Reviewer 3: It is unfortunate that the `render()` method is explained that far in the article. It seems to be the key subject of this article, as it tells how to define the content that developers could then change dynamically with bindings, style, listen events on, etc...

I would rather move this section to be first in the article, so that it contains the first code example, and tells how to register a component and define its HTML content.

Author: Fixed
